batch/batch/worker/worker.py

```diff
@@ -2942,39 +2983,54 @@ def __init__(self, client_session: httpx.ClientSession):
 
         self.cloudfuse_mount_manager = ReadOnlyCloudfuseManager()
 
+        self._jvmpools_by_cores: Dict[int, JVMQueue] = {
+            n_cores: JVMPool(n_cores) for n_cores in (1, 2, 4, 8)
+        }
+        self._waiting_for_jvm_with_n_cores: asyncio.Queue[int] = asyncio.Queue()
         self._jvm_initializer_task = asyncio.create_task(self._initialize_jvms())
-        self._jvms = SortedSet([], key=lambda jvm: jvm.n_cores)
 
     async def _initialize_jvms(self):
         assert instance_config
-        if instance_config.worker_type() in ('standard', 'D', 'highmem', 'E'):
-            jvms: List[Awaitable[JVM]] = []
-            for jvm_cores in (1, 2, 4, 8):
-                for _ in range(CORES // jvm_cores):
-                    jvms.append(JVM.create(len(jvms), jvm_cores, self))
-            assert len(jvms) == N_JVM_CONTAINERS
-            self._jvms.update(await asyncio.gather(*jvms))
-        log.info(f'JVMs initialized {self._jvms}')
+        if instance_config.worker_type() not in ('standard', 'D', 'highmem', 'E'):
+            log.info(f'no JVMs initialized')
+
+        while True:
+            try:
+                n_cores = self._waiting_for_jvm_with_n_cores.get_nowait()
+                await self._jvmpools_by_cores[n_cores].create_jvm()
+            except asyncio.QueueEmpty:
+                next_unfull_jvmpool = None
+                for jvmpool in self._jvmpools_by_cores.values():
+                    if not jvmpool.full():
+                        next_unfull_jvmpool = jvmpool
+                        break
+
+                if next_unfull_jvmpool is None:
+                    break
+                await next_unfull_jvmpool.create_jvm()
+
+        assert self._waiting_for_jvm_with_n_cores.empty()
+        assert all(jvmpool.full() for jvmpool in self._jvmpools_by_cores.values())
+        log.info(f'JVMs initialized {self._jvmpools_by_cores}')
```

Comment:
@jigold Thanks for pushing back! What do you think of it now?

To directly answer your question: one queue (`jvmpool.queue`) is a place for a consumer to borrow a JVM, the other queue (`waiting_for_jvm_with_n_cores`) is a place for a producer to learn that a consumer is waiting.

Without `waiting_for_jvm_with_n_cores`, `_initialize_jvms` has no way to be told that someone is waiting for a JVM. `asyncio.Queue` doesn't expose a method like `has_waiters()`.